components/content-service/pkg/storage/s3.go

```diff
@@ -0,0 +1,417 @@
+	resp, err := rs.client.GetObjectAttributes(ctx, &s3.GetObjectAttributesInput{
+		Bucket:           &rs.Config.Bucket,
+		Key:              aws.String(obj),
+		ObjectAttributes: []types.ObjectAttributes{types.ObjectAttributesObjectSize},
+	})
+
+	var nsk *types.NoSuchKey
+	if errors.As(err, &nsk) {
+		return nil, ErrNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := rs.PresignedFactory().PresignGetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(rs.Config.Bucket),
+		Key:    aws.String(obj),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &DownloadInfo{
+		Meta: ObjectMeta{
+			// TODO(cw): implement this if we need to support FWB with S3
+		},
+		Size: resp.ObjectSize,
+		URL:  req.URL,
+	}, nil
+}
+
+// SignUpload implements PresignedAccess
+func (rs *PresignedS3Storage) SignUpload(ctx context.Context, bucket string, obj string, options *SignedURLOptions) (info *UploadInfo, err error) {
+	resp, err := rs.PresignedFactory().PresignPutObject(ctx, &s3.PutObjectInput{
+		Bucket: &rs.Config.Bucket,
+		Key:    aws.String(obj),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &UploadInfo{
+		URL: resp.URL,
+	}, nil
+}
+
+func newDirectS3Access(client S3Client, config S3Config) *s3Storage {
+	return &s3Storage{
+		Config: config,
+		client: client,
+	}
+}
+
+type s3Storage struct {
+	Config S3Config
+
+	OwnerID, WorkspaceID, InstanceID string
+
+	client S3Client
+}
+
+// Bucket implements DirectAccess
+func (s3st *s3Storage) Bucket(userID string) string {
+	return s3st.Config.Bucket
+}
+
+// BackupObject implements DirectAccess
+func (s3st *s3Storage) BackupObject(name string) string {
+	return s3st.objectName(name)
+}
+
+// Download implements DirectAccess
+func (s3st *s3Storage) Download(ctx context.Context, destination string, name string, mappings []archive.IDMapping) (found bool, err error) {
+	return s3st.download(ctx, destination, s3st.objectName(name), mappings)
+}
+
+// DownloadSnapshot implements DirectAccess
+func (s3st *s3Storage) DownloadSnapshot(ctx context.Context, destination string, name string, mappings []archive.IDMapping) (found bool, err error) {
+	return s3st.download(ctx, destination, name, mappings)
+}
+
+func (s3st *s3Storage) download(ctx context.Context, destination string, obj string, mappings []archive.IDMapping) (found bool, err error) {
+	resp, err := s3st.client.GetObject(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(s3st.Config.Bucket),
+		Key:    aws.String(obj),
+	})
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	err = archive.ExtractTarbal(ctx, resp.Body, destination, archive.WithUIDMapping(mappings), archive.WithGIDMapping(mappings))
+	if err != nil {
+		return true, xerrors.Errorf("tar %s: %s", destination, err.Error())
+	}
+
+	return true, nil
+}
+
+// EnsureExists implements DirectAccess
+func (*s3Storage) EnsureExists(ctx context.Context) error {
+	return nil
```

Comment:
Why we did not implement this?

Reply:
According to the comment https://github.com/gitpod-io/gitpod/blob/8b75aba6abfa82d98ff4575a4f05847989326697/install/installer/pkg/config/v1/config.go#L261-L262
Looks like we fall back to "one bucket per user" behavior.

Or, no? We only have a single bucket behavior now. So we have to change the [BucketName](https://github.com/gitpod-io/gitpod/blob/8b75aba6abfa82d98ff4575a4f05847989326697/install/installer/pkg/config/v1/config.go#L263) as a required field?

Reply:
That comment is not accurate anymore. I will change the field to required.